### ISSUE TITLE
system testing and reusing test sections

### ISSUE THREAD

**Issue by adrian-herscu** [TNG/JGiven#587]
I am using JGiven on top of TestNG for about 6 years and it is a great framework :)

Besides some headaches I had with running tests in parallel, for which I opened several issues here, I also found it difficult to deal with multiple interfaces. In systems testing it is common that a test is required to access the system via two or more interfaces. For example, in a simple client-server system, a test might need to access the client via Selenium and the server via REST. There are even more complex scenarios which require SSH, JDBC, Kafka, RabitMQ, CloudFoundry, ElasticSearch... In current system we have about 200 tests! What a headache...

So, what is the problem? JGiven is designed to have a 3 stage test scenario: given, when, then. Each of these stages are modeled as classes holding step methods to be referenced by various test scenarios. This way, for a Selenium only scenario, its when steps would be clicking, typing, etc. But what happens if the real scenario also requires REST? First thought might be "let's add REST steps" to that class. Then you discover that the imports list of that class grows larger and larger an so its methods number. My solution is to hold separate sets of given/when/then classes for each kind of interface. The test scenario now has to hold additional `@ScenarioStage` fields and reference these like `restActions.when().posting(...)`.  Not a big issue. Further test development arise the need to reuse common scenario sections. For example, in a multi-user application it is imperative to have some sort of authentication, so now each test scenario must begin with a login section. One might say that login can or should be modeled as a step method. However, in a real test you might be required to verify for successful login. So you cannot model that as a step now, otherwise you would mix actions with verifications, undermining the whole concept of having them separated as given/when/then from the beginning. I solved this by creating a common base class for all test scenario classes just to hold these section methods and as bonus also to hold these additional `@ScenarioStage` fields and also to initialize the various clients needed, Selenium, REST & Co. This keeps the scenario test classes clean of any infrastructure code.

So far so good. But yet there is another issue which I could not find an easy way to solve :(

Some test scenarios do not require the full-stack of interfaces. Let's say they only require REST. Then initializing a Selenium client for them is both time and resource consuming. Hence these scenarios cannot inherit from same base class that holds our common section methods. As a temporary solution we duplicated the common section methods. But that list also grows larger from release to release.

My first attempt was to create a common sections class and reuse it via containment and delegation from other base classes. It's constructor requires the given/when/then stages and a configuration object in order to hold for its sections. From other base classes I need to initialize this common class. So first I tried to do that in TestNG `@BeforeClass` method. Didn't work. Tried with a `@BeforeMethod`. I am getting NPE for one of my additional stages.

Just to summarize:
1. How can hook in my initialization before a test method begins and after its stages are initialized?
2. Any one her has a simpler way to deal with this mess?

**Comment by adrian-herscu** [comment]
Solved by introducing another base class that does not initialize the Selenium client. Still, there might be similar situations with other types of clients.
